File: open_spiel/python/algorithms/cfr_test.py
# Copyright 2019 DeepMind Technologies Limited
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Tests for open_spiel.python.algorithms.cfr."""

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import itertools

from absl.testing import absltest
from absl.testing import parameterized
import numpy as np

from open_spiel.python import policy
from open_spiel.python.algorithms import cfr
from open_spiel.python.algorithms import expected_game_score
from open_spiel.python.algorithms import exploitability
import pyspiel

_KUHN_GAME = pyspiel.load_game("kuhn_poker")
_LEDUC_GAME = pyspiel.load_game("leduc_poker")

_KUHN_UNIFORM_POLICY = policy.TabularPolicy(_KUHN_GAME)
_LEDUC_UNIFORM_POLICY = policy.TabularPolicy(_LEDUC_GAME)


class ModuleLevelFunctionTest(absltest.TestCase):

  def test__update_current_policy(self):
    game = pyspiel.load_game("kuhn_poker")
    tabular_policy = policy.TabularPolicy(game)

    cumulative_regrets = np.arange(0, 12 * 2).reshape((12, 2))
    expected_policy = cumulative_regrets / np.sum(
        cumulative_regrets, axis=-1, keepdims=True)
    nodes_indices = {
        u"0": 0,
        u"0pb": 1,
        u"1": 2,
        u"1pb": 3,
        u"2": 4,
        u"2pb": 5,
        u"1p": 6,
        u"1b": 7,
        u"2p": 8,
        u"2b": 9,
        u"0p": 10,
        u"0b": 11,
    }
    # pylint: disable=g-complex-comprehension
    info_state_nodes = {
        key: cfr._InfoStateNode(
            legal_actions=[0, 1],
            index_in_tabular_policy=None,
            cumulative_regret=dict(enumerate(cumulative_regrets[index])),
            cumulative_policy=None) for key, index in nodes_indices.items()
    }
    # pylint: enable=g-complex-comprehension

    cfr._update_current_policy(tabular_policy, info_state_nodes)

    np.testing.assert_array_equal(expected_policy,
                                  tabular_policy.action_probability_array)


class CFRTest(parameterized.TestCase, absltest.TestCase):

  @parameterized.parameters(
      list(itertools.product([True, False], [True, False], [True, False])))
  def test_policy_zero_is_uniform(self, linear_averaging, regret_matching_plus,
                                  alternating_updates):
    # We use Leduc and not Kuhn, because Leduc has illegal actions and Kuhn does
    # not.
    game = pyspiel.load_game("leduc_poker")
    cfr_solver = cfr._CFRSolver(
        game,
        regret_matching_plus=regret_matching_plus,
        linear_averaging=linear_averaging,
        alternating_updates=alternating_updates)

    np.testing.assert_array_equal(
        _LEDUC_UNIFORM_POLICY.action_probability_array,
        cfr_solver.current_policy().action_probability_array)
    np.testing.assert_array_equal(
        _LEDUC_UNIFORM_POLICY.action_probability_array,
        cfr_solver.average_policy().action_probability_array)

  def test_cfr_kuhn_poker(self):
    game = pyspiel.load_game("kuhn_poker")
    cfr_solver = cfr.CFRSolver(game)
    for _ in range(300):
      cfr_solver.evaluate_and_update_policy()
    average_policy = cfr_solver.average_policy()
    average_policy_values = expected_game_score.policy_value(
        game.new_initial_state(), [average_policy] * 2)
    # 1/18 is the Nash value. See https://en.wikipedia.org/wiki/Kuhn_poker
    np.testing.assert_allclose(
        average_policy_values, [-1 / 18, 1 / 18], atol=1e-3)

  def test_cfr_plus_kuhn_poker(self):
    game = pyspiel.load_game("kuhn_poker")
    cfr_solver = cfr.CFRPlusSolver(game)
    for _ in range(200):
      cfr_solver.evaluate_and_update_policy()
    average_policy = cfr_solver.average_policy()
    average_policy_values = expected_game_score.policy_value(
        game.new_initial_state(), [average_policy] * 2)
    # 1/18 is the Nash value. See https://en.wikipedia.org/wiki/Kuhn_poker
    np.testing.assert_allclose(
        average_policy_values, [-1 / 18, 1 / 18], atol=1e-3)

  def test_cfr_plus_solver_best_response_mdp(self):
    game = pyspiel.load_game("kuhn_poker")
    cfr_solver = cfr.CFRPlusSolver(game)
    for _ in range(200):
      cfr_solver.evaluate_and_update_policy()
    average_policy = cfr_solver.average_policy()
    pyspiel_avg_policy = policy.python_policy_to_pyspiel_policy(average_policy)
    br_computer = pyspiel.TabularBestResponseMDP(game, pyspiel_avg_policy)
    br_info = br_computer.exploitability()
    self.assertLessEqual(br_info.exploitability, 0.001)

  def test_cfr_cce_ce_dist_goofspiel(self):
    """Copy of the TestCCEDistCFRGoofSpiel in corr_dist_test.cc."""
    game = pyspiel.load_game(
        "turn_based_simultaneous_game(game=goofspiel(num_cards=3,points_order="
        "descending,returns_type=total_points))")
    for num_iterations in [1, 10, 100]:
      policies = []
      cfr_solver = cfr.CFRSolver(game)
      for _ in range(num_iterations):
        cfr_solver.evaluate_and_update_policy()
        policies.append(
            policy.python_policy_to_pyspiel_policy(cfr_solver.current_policy()))
      mu = pyspiel.uniform_correlation_device(policies)
      cce_dist_info = pyspiel.cce_dist(game, mu)
      print("goofspiel, cce test num_iters: {}, cce_dist: {}, per player: {}"
            .format(num_iterations, cce_dist_info.dist_value,
                    cce_dist_info.deviation_incentives))
      # Try converting one of the BR policies:
      _ = policy.pyspiel_policy_to_python_policy(
          game, cce_dist_info.best_response_policies[0])

      # Assemble the same correlation device manually, just as an example for
      # how to do non-uniform distributions of them and to test the python
      # bindings for lists of tuples works properly
      uniform_prob = 1.0 / len(policies)
      mu2 = [(uniform_prob, policy) for policy in policies]
      cce_dist_info2 = pyspiel.cce_dist(game, mu2)
      self.assertAlmostEqual(cce_dist_info2.dist_value,
                             sum(cce_dist_info.deviation_incentives))
      # Test the CEDist function too, why not. Disable the exact one, as it
      # takes too long for a test.
      # ce_dist_info = pyspiel.ce_dist(game, pyspiel.determinize_corr_dev(mu))
      ce_dist_info = pyspiel.ce_dist(
          game, pyspiel.sampled_determinize_corr_dev(mu, 100))
      print("goofspiel, ce test num_iters: {}, ce_dist: {}, per player: {}"
            .format(num_iterations, ce_dist_info.dist_value,
                    ce_dist_info.deviation_incentives))
      print("number of conditional best responses per player:")
      for p in range(game.num_players()):
        print("  player {}, num: {}".format(
            p, len(ce_dist_info.conditional_best_response_policies[p])))

  @parameterized.parameters(
      list(itertools.product([True, False], [True, False], [True, False])))
  def test_cfr_kuhn_poker_runs_with_multiple_players(self, linear_averaging,
                                                     regret_matching_plus,
                                                     alternating_updates):
    num_players = 3

    game = pyspiel.load_game("kuhn_poker", {"players": num_players})
    cfr_solver = cfr._CFRSolver(
        game,
        regret_matching_plus=regret_matching_plus,
        linear_averaging=linear_averaging,
        alternating_updates=alternating_updates)
    for _ in range(10):
      cfr_solver.evaluate_and_update_policy()
    average_policy = cfr_solver.average_policy()
    average_policy_values = expected_game_score.policy_value(
        game.new_initial_state(), [average_policy] * num_players)
    del average_policy_values

  @parameterized.parameters(list(itertools.product([False, True])))
  def test_simultaneous_two_step_avg_1b_seq_in_kuhn_poker(
      self, regret_matching_plus):
    num_players = 2
    game = pyspiel.load_game("kuhn_poker", {"players": num_players})
    cfr_solver = cfr._CFRSolver(
        game,
        regret_matching_plus=regret_matching_plus,
        linear_averaging=False,
        alternating_updates=False)

    def check_avg_policy_is_uniform_random():
      avg_policy = cfr_solver.average_policy()
      for player_info_states in avg_policy.states_per_player:
        for info_state in player_info_states:
          state_policy = avg_policy.policy_for_key(info_state)
          np.testing.assert_allclose(state_policy, [1.0 / len(state_policy)] *
                                     len(state_policy))

    check_avg_policy_is_uniform_random()

    cfr_solver.evaluate_and_update_policy()
    check_avg_policy_is_uniform_random()

    cfr_solver.evaluate_and_update_policy()

    # The acting player in 1b is player 1 and they have not acted before, so
    # the probability this player plays to this information state is 1, and
    # the sequence probability of any action is just the probability of that
    # action given the information state. On the first iteration, this
    # probability is 0.5 for both actions. On the second iteration, the
    # current policy is [0, 1], so the average cumulants should be
    # [0.5, 1.5]. Normalizing this gives the average policy.
    normalization = 0.5 + 0.5 + 1
    np.testing.assert_allclose(cfr_solver.average_policy().policy_for_key("1b"),
                               [0.5 / normalization, (0.5 + 1) / normalization])

  def test_policy(self):
    game = pyspiel.load_game("kuhn_poker")
    solver = cfr.CFRPlusSolver(game)

    tabular_policy = solver.current_policy()
    self.assertLen(tabular_policy.state_lookup, 12)
    for info_state_str in tabular_policy.state_lookup.keys():
      np.testing.assert_equal(
          np.asarray([0.5, 0.5]), tabular_policy.policy_for_key(info_state_str))

  @parameterized.parameters([
      (pyspiel.load_game("kuhn_poker"), pyspiel.CFRSolver, cfr.CFRSolver),
      (pyspiel.load_game("leduc_poker"), pyspiel.CFRSolver, cfr.CFRSolver),
      (pyspiel.load_game("kuhn_poker"), pyspiel.CFRPlusSolver,
       cfr.CFRPlusSolver),
      (pyspiel.load_game("leduc_poker"), pyspiel.CFRPlusSolver,
       cfr.CFRPlusSolver),
  ])
  def test_cpp_algorithms_identical_to_python_algorithm(self, game, cpp_class,
                                                        python_class):
    cpp_solver = cpp_class(game)
    python_solver = python_class(game)

    for _ in range(5):
      cpp_solver.evaluate_and_update_policy()
      python_solver.evaluate_and_update_policy()

      cpp_avg_policy = cpp_solver.average_policy()
      python_avg_policy = python_solver.average_policy()

      # We do not compare the policy directly as we do not have an easy way to
      # convert one to the other, so we use the exploitability as a proxy.
      cpp_expl = pyspiel.nash_conv(game, cpp_avg_policy)
      python_expl = exploitability.nash_conv(game, python_avg_policy)
      self.assertEqual(cpp_expl, python_expl)
    # Then we also check the CurrentPolicy, just to check it is giving the same
    # results too
    cpp_current_policy = cpp_solver.current_policy()
    python_current_policy = python_solver.current_policy()
    cpp_expl = pyspiel.nash_conv(game, cpp_current_policy)
    python_expl = exploitability.nash_conv(game, python_current_policy)
    self.assertEqual(cpp_expl, python_expl)


if __name__ == "__main__":
  absltest.main()

File: open_spiel/algorithms/cfr_test.cc
// Copyright 2021 DeepMind Technologies Limited
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "open_spiel/algorithms/cfr.h"

#include <cmath>
#include <iostream>

#include "open_spiel/algorithms/expected_returns.h"
#include "open_spiel/algorithms/history_tree.h"
#include "open_spiel/algorithms/tabular_exploitability.h"
#include "open_spiel/game_transforms/turn_based_simultaneous_game.h"
#include "open_spiel/games/kuhn_poker.h"
#include "open_spiel/games/leduc_poker.h"
#include "open_spiel/games/liars_dice.h"
#include "open_spiel/games/matching_pennies_3p.h"
#include "open_spiel/games/tic_tac_toe.h"
#include "open_spiel/spiel.h"
#include "open_spiel/spiel_utils.h"

namespace open_spiel {
namespace algorithms {
namespace {

void CheckNashKuhnPoker(const Game& game, const Policy& policy) {
  const std::vector<double> game_value =
      ExpectedReturns(*game.NewInitialState(), policy, -1);

  // 1/18 is the Nash value. See https://en.wikipedia.org/wiki/Kuhn_poker
  constexpr float nash_value = 1.0 / 18.0;
  constexpr float eps = 1e-3;

  SPIEL_CHECK_EQ(2, game_value.size());
  SPIEL_CHECK_FLOAT_NEAR((float)game_value[0], -nash_value, eps);
  SPIEL_CHECK_FLOAT_NEAR((float)game_value[1], nash_value, eps);
}

void CheckExploitabilityKuhnPoker(const Game& game, const Policy& policy) {
  SPIEL_CHECK_LE(Exploitability(game, policy), 0.05);
}

void CFRTest_KuhnPoker() {
  std::shared_ptr<const Game> game = LoadGame("kuhn_poker");
  CFRSolver solver(*game);
  for (int i = 0; i < 300; i++) {
    solver.EvaluateAndUpdatePolicy();
  }
  const std::shared_ptr<Policy> average_policy = solver.AveragePolicy();
  CheckNashKuhnPoker(*game, *average_policy);
  CheckExploitabilityKuhnPoker(*game, *average_policy);
}

void CFRTest_IIGoof4() {
  // Random points order.
  std::shared_ptr<const Game> game = LoadGameAsTurnBased(
      "goofspiel", {{"imp_info", GameParameter(true)},
                    {"points_order", GameParameter(std::string("random"))},
                    {"num_cards", GameParameter(4)}});

  CFRSolver solver(*game);
  for (int i = 0; i < 100; i++) {
    solver.EvaluateAndUpdatePolicy();
  }
  // Values checked with Marc's thesis implementation.
  const std::shared_ptr<Policy> average_policy = solver.AveragePolicy();
  SPIEL_CHECK_LE(Exploitability(*game, *average_policy), 0.1);

  // Fixed points order.
  game = LoadGameAsTurnBased(
      "goofspiel", {{"imp_info", GameParameter(true)},
                    {"points_order", GameParameter(std::string("descending"))},
                    {"num_cards", GameParameter(4)}});

  CFRSolver solver2(*game);
  for (int i = 0; i < 1000; i++) {
    solver2.EvaluateAndUpdatePolicy();
  }
  // Values checkes with Marc's thesis implementation.
  const std::shared_ptr<Policy> average_policy2 = solver2.AveragePolicy();
  SPIEL_CHECK_LE(Exploitability(*game, *average_policy2), 0.01);
}

void CFRPlusTest_KuhnPoker() {
  std::shared_ptr<const Game> game = LoadGame("kuhn_poker");
  CFRPlusSolver solver(*game);
  for (int i = 0; i < 200; i++) {
    solver.EvaluateAndUpdatePolicy();
  }
  const std::shared_ptr<Policy> average_policy = solver.AveragePolicy();
  CheckNashKuhnPoker(*game, *average_policy);
  CheckExploitabilityKuhnPoker(*game, *average_policy);
}

void CFRTest_KuhnPokerRunsWithThreePlayers(bool linear_averaging,
                                           bool regret_matching_plus,
                                           bool alternating_updates) {
  int num_players = 3;
  std::shared_ptr<const Game> game =
      LoadGame("kuhn_poker", {{"players", GameParameter(num_players)}});
  CFRSolverBase solver(*game, alternating_updates, linear_averaging,
                       regret_matching_plus);
  for (int i = 0; i < 10; i++) {
    solver.EvaluateAndUpdatePolicy();
  }
  std::shared_ptr<Policy> average_policy = solver.AveragePolicy();
  // Value upper-bounds inspired by Fig 2 of (Srinivasan et al., Actor-Critic
  // Policy Optimization in Partially Observable Multiagent Environments, 2018)
  // https://arxiv.org/abs/1810.09026
  SPIEL_CHECK_LE(NashConv(*game, *average_policy), 1);
}

// Tests the convergence of CFR in a specific game. Only computes nashconv
// if the upper bound is positive.
void CFRTest_GeneralMultiplePlayerTest(const std::string& game_name,
                                       int num_players, int num_iterations,
                                       double nashconv_upper_bound) {
  std::shared_ptr<const Game> game =
      LoadGame(game_name, {{"players", GameParameter(num_players)}});
  CFRSolverBase solver(*game,
                       /*alternating_updates=*/true,
                       /*linear_averaging=*/false,
                       /*regret_matching_plus=*/false);
  for (int i = 0; i < num_iterations; i++) {
    solver.EvaluateAndUpdatePolicy();
  }

  if (nashconv_upper_bound > 0) {
    std::shared_ptr<Policy> average_policy = solver.AveragePolicy();
    SPIEL_CHECK_LE(NashConv(*game, *average_policy), nashconv_upper_bound);
  }
}

void CFRTest_OneShotGameTest(int iterations, std::string one_shot_game,
                             double nashconv_upper_bound) {
  // Note: this is a 3-player general sum game with a unique uniform mixed
  // strategy. However, CFR is not guaranteed to converge, and indeed fails to,
  // just like was shown for fictitious play.
  std::cout << one_shot_game << " convergence test." << std::endl;
  std::shared_ptr<const Game> game = LoadGameAsTurnBased(one_shot_game);
  CFRSolverBase solver(*game,
                       /*alternating_updates=*/true,
                       /*linear_averaging=*/false,
                       /*regret_matching_plus=*/false);
  double nash_conv = 0;
  for (int i = 0; i < iterations; i++) {
    solver.EvaluateAndUpdatePolicy();
    if (i % 10 == 0) {
      std::shared_ptr<Policy> average_policy = solver.AveragePolicy();
      nash_conv = NashConv(*game, *average_policy);
      std::cout << "iter " << i << ", nashconv = " << nash_conv << std::endl;

      if (game->GetType().utility == GameType::Utility::kConstantSum ||
          game->GetType().utility == GameType::Utility::kZeroSum) {
        double expl = Exploitability(*game, *average_policy);
        SPIEL_CHECK_FLOAT_NEAR(expl, nash_conv / game->NumPlayers(), 1e-10);
      }
    }
  }
  SPIEL_CHECK_LE(nash_conv, nashconv_upper_bound);
}

// Tests the convergence of CFR in a specific game. Only computes nashconv
// if the upper bound is positive.
void CFRTest_TicTacToe(int num_iterations, double nashconv_upper_bound) {
  std::shared_ptr<const Game> game = LoadGame("tic_tac_toe");
  CFRSolverBase solver(*game,
                       /*alternating_updates=*/true,
                       /*linear_averaging=*/false,
                       /*regret_matching_plus=*/false);
  for (int i = 0; i < num_iterations; i++) {
    solver.EvaluateAndUpdatePolicy();
  }

  if (nashconv_upper_bound > 0) {
    std::shared_ptr<Policy> average_policy = solver.AveragePolicy();
    SPIEL_CHECK_LE(NashConv(*game, *average_policy), nashconv_upper_bound);
  }
}

void CFRTest_InfoStateValuesTableSerialization() {
  // Check empty
  CFRInfoStateValuesTable info_state_values_table = {};
  std::string serialized0 = "";
  SerializeCFRInfoStateValuesTable(info_state_values_table, &serialized0, -1);
  CFRInfoStateValuesTable deserialized0;
  DeserializeCFRInfoStateValuesTable(serialized0, &deserialized0);
  SPIEL_CHECK_TRUE(deserialized0.empty());

  // Check non-empty
  info_state_values_table = {
      {"", CFRInfoStateValues({0}, 1.0)},
      {"0:0,0;0", CFRInfoStateValues({0, 1, 2}, 0.1)},
      {"<->\n<->", CFRInfoStateValues({0, 1, 2}, 0.1)},
      {"1:1,1;1", CFRInfoStateValues({0, 1, 2, 3}, 0.2)}};
  std::string serialized1 = "";
  SerializeCFRInfoStateValuesTable(info_state_values_table, &serialized1, -1);
  CFRInfoStateValuesTable deserialized1;
  DeserializeCFRInfoStateValuesTable(serialized1, &deserialized1);

  SPIEL_CHECK_EQ(info_state_values_table.size(),
                 info_state_values_table.size());
  for (const auto& [info_state, values] : info_state_values_table) {
    for (int i = 0; i < values.legal_actions.size(); i++) {
      SPIEL_CHECK_EQ(values.legal_actions.at(i),
                     deserialized1.at(info_state).legal_actions.at(i));
      SPIEL_CHECK_FLOAT_NEAR(
          values.cumulative_regrets.at(i),
          deserialized1.at(info_state).cumulative_regrets.at(i), 1e-15);
      SPIEL_CHECK_FLOAT_NEAR(
          values.cumulative_policy.at(i),
          deserialized1.at(info_state).cumulative_policy.at(i), 1e-15);
      SPIEL_CHECK_FLOAT_NEAR(values.current_policy.at(i),
                             deserialized1.at(info_state).current_policy.at(i),
                             1e-15);
    }
  }
}

void CFRTest_CFRSolverSerialization() {
  auto game = LoadGame("kuhn_poker");
  CFRSolver solver = CFRSolver(*game);
  double exploitability0 = Exploitability(*game, *solver.AveragePolicy());

  for (int i = 0; i < 50; i++) {
    solver.EvaluateAndUpdatePolicy();
  }
  double exploitability1 = Exploitability(*game, *solver.AveragePolicy());
  SPIEL_CHECK_GT(exploitability0, exploitability1);

  std::string serialized = solver.Serialize();
  std::unique_ptr<CFRSolver> deserialized_solver =
      DeserializeCFRSolver(serialized);
  SPIEL_CHECK_EQ(solver.InfoStateValuesTable().size(),
                 deserialized_solver->InfoStateValuesTable().size());
  double exploitability2 =
      Exploitability(*game, *deserialized_solver->AveragePolicy());
  SPIEL_CHECK_FLOAT_NEAR(exploitability1, exploitability2, 1e-15);

  for (int i = 0; i < 50; i++) {
    deserialized_solver->EvaluateAndUpdatePolicy();
  }
  double exploitability3 =
      Exploitability(*game, *deserialized_solver->AveragePolicy());
  SPIEL_CHECK_GT(exploitability2, exploitability3);
}

}  // namespace
}  // namespace algorithms
}  // namespace open_spiel

namespace algorithms = open_spiel::algorithms;

int main(int argc, char** argv) {
  algorithms::CFRTest_KuhnPoker();
  algorithms::CFRTest_IIGoof4();
  algorithms::CFRPlusTest_KuhnPoker();
  algorithms::CFRTest_KuhnPokerRunsWithThreePlayers(
      /*linear_averaging=*/false,
      /*regret_matching_plus=*/false,
      /*alternating_updates=*/false);
  algorithms::CFRTest_KuhnPokerRunsWithThreePlayers(
      /*linear_averaging=*/true,
      /*regret_matching_plus=*/false,
      /*alternating_updates=*/false);
  algorithms::CFRTest_KuhnPokerRunsWithThreePlayers(
      /*linear_averaging=*/true,
      /*regret_matching_plus=*/true,
      /*alternating_updates=*/false);
  algorithms::CFRTest_KuhnPokerRunsWithThreePlayers(
      /*linear_averaging=*/true,
      /*regret_matching_plus=*/true,
      /*alternating_updates=*/true);

  // Value upper bounds for Kuhn taken from Figure 2 of (Lanctot, Further
  // Developments of Extensive-Form Replicator Dynamics using the Sequence-Form
  // Representation, 2014).
  algorithms::CFRTest_GeneralMultiplePlayerTest(
      /*game_name=*/"kuhn_poker", /*num_players=*/3, /*num_iterations=*/10,
      /*nashconv_upper_bound=*/1.0);
  algorithms::CFRTest_GeneralMultiplePlayerTest(
      /*game_name=*/"kuhn_poker", /*num_players=*/4, /*num_iterations=*/10,
      /*nashconv_upper_bound=*/1.0);

  // Value upper-bounds for Leduc taken from Figure 17 of (Lanctot et al.,
  // A Unified Game-Theoretic Approach to Multiagent Reinforcement Learning,
  // 2017). See https://arxiv.org/abs/1711.00832. For Liar's Dice, number's were
  // obtained using Marc's thesis code.
  algorithms::CFRTest_GeneralMultiplePlayerTest(
      /*game_name=*/"leduc_poker", /*num_players=*/2, /*num_iterations=*/10,
      /*nashconv_upper_bound=*/2.0);

  // Disabled as they make the test too long for a unit test.
  // algorithms::CFRTest_GeneralMultiplePlayerTest(
  //    /*game_name=*/"liars_dice", /*num_players=*/2, /*num_iterations=*/10,
  //    /*nashconv_upper_bound=*/1.0);
  // algorithms::CFRTest_GeneralMultiplePlayerTest(
  //   /*game_name=*/"leduc_poker", /*num_players=*/3, /*num_iterations=*/2,
  //   /*nashconv_upper_bound=*/10.0);

  // Test a few one-shot games.
  algorithms::CFRTest_OneShotGameTest(1000, "matrix_rps", 1e-6);
  algorithms::CFRTest_OneShotGameTest(1000, "matrix_shapleys_game", 1.0);
  algorithms::CFRTest_OneShotGameTest(1000, "matching_pennies_3p", 3.0);

  // Try out a perfect information game that has an InformationState that obeys
  // perfect recall. Test is disabled because it adds 20 seconds. Let's revisit
  // when we add a version that can handle safe imperfect recall information
  // states.
  // algorithms::CFRTest_TicTacToe(10, 2.0);

  algorithms::CFRTest_InfoStateValuesTableSerialization();
  algorithms::CFRTest_CFRSolverSerialization();
}
